Src/Particle/AMReX_ParticleIO.H

```diff
@@ -400,302 +402,19 @@ ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt>
-                       1,0,info);
-        
-        // We eventually want to write out the file name and the offset
-        // into that file into which each grid of particles is written.
-        Vector<int>  which(state.size(),0);
-        Vector<int > count(state.size(),0);
-        Vector<Long> where(state.size(),0);
-	
-        std::string filePrefix(LevelDir);
-        filePrefix += '/';
-        filePrefix += ParticleType::DataPrefix();
-        if(usePrePost) {
-            filePrefixPrePost[lev] = filePrefix;
-        }
-        bool groupSets(false), setBuf(true);
-        
-        if (gotsome)
-        {
-            for(NFilesIter nfi(nOutFiles, filePrefix, groupSets, setBuf); nfi.ReadyToWrite(); ++nfi)
-            {
-                std::ofstream& myStream = (std::ofstream&) nfi.Stream();
-                WriteParticles(lev, myStream, nfi.FileNumber(), which, count, where,
-                               write_real_comp, write_int_comp, particle_io_flags);
-            }
-            
-            if(usePrePost) {
-                whichPrePost[lev] = which;
-                countPrePost[lev] = count;
-                wherePrePost[lev] = where;
-            } else {
-                ParallelDescriptor::ReduceIntSum (which.dataPtr(), which.size(), IOProcNumber);
-                ParallelDescriptor::ReduceIntSum (count.dataPtr(), count.size(), IOProcNumber);
-                ParallelDescriptor::ReduceLongSum(where.dataPtr(), where.size(), IOProcNumber);
-            }
-        }
-	
-        if (ParallelDescriptor::IOProcessor())
-        {
-            if(usePrePost) {
-                // ---- write to the header and unlink in CheckpointPost
-            } else {
-                for (int j = 0; j < state.size(); j++)
-                {
-                    HdrFile << which[j] << ' ' << count[j] << ' ' << where[j] << '\n';
-                }
-				
-                if (gotsome && doUnlink)
-                {
-                    // Unlink any zero-length data files.
-                    Vector<Long> cnt(nOutFiles,0);
-					
-                    for (int i = 0, N=count.size(); i < N; i++) {
-                        cnt[which[i]] += count[i];
-                    }
-                    
-                    for (int i = 0, N=cnt.size(); i < N; i++)
-                    {
-                        if (cnt[i] == 0)
-                        {
-                            std::string FullFileName = NFilesIter::FileName(i, filePrefix);
-                            FileSystem::Remove(FullFileName);
-                        }
-                    }
-                }                
-            }            
-        }
-    }
-    
-    if (ParallelDescriptor::IOProcessor())
+    if (AsyncOut::UseAsyncOut()) {
+        WriteBinaryParticleDataAsync(*this, dir, name,
+                                     write_real_comp, write_int_comp,
+                                     real_comp_names, int_comp_names);
```

Comment:
Any reason why the async version does not take `F`?

Reply:
I plan to remove `F` soon. WarpX will implement this differently, by copying to a tmp particle container and applying a filter during the copy. That functionality will be useful in other places, and it will let us remove all those extra overloads of `WritePlotFile` and `Checkpoint`.